### DISCUSSION TITLE
Is there a possibility for paginate the collection?

### DISCUSSION THREAD

**Discussion by ElisameAraujo** [staudenmeir/laravel-adjacency-list#227] [Q&A] [answered]
The question is quite simple: 

I have a collection of `Categorias` with relantionship with `Subcategorias`. 

How I can paginate this categories in chunks similar to Laravel pagination? Altough I can use `Categorias::tree()->paginate(2)->toTree()`, I can't render the method `links` using like this `{{ $categorias->links('vendor/pagination/bs-5')}}`.

Is there a option to paginate or not? If it is, how I can do? If not, is possible to add this option in the near future?
![pagination](https://github.com/staudenmeir/laravel-adjacency-list/assets/40779799/bda19146-31fe-4fd8-85e5-9b22da0e4015)

**Comment by staudenmeir** [comment, marked as answer]
Hi @ElisameAraujo,
Did you try `Categorias::tree()->paginate(2)` (without `->toTree()`)?

It's possible to make pagination work with `Categorias::tree()->paginate(2)->toTree()`, but that only gives you the root categories and I assume that's not what you want.

**Comment by ElisameAraujo** [comment]
For the first answer, no, but I can try it.

For the second, yep, is not what I want 
I want paginate categories with child categories, some like WordPress do it.
Actually, the structure the list my categories with his subs is:

![image](https://github.com/staudenmeir/laravel-adjacency-list/assets/40779799/d1f23571-634b-4efb-adf6-911c1c2db994)

Where **Categoria Principal** is the parent category and the **Nome da Categoria** with a dash (&mdash;) is a child under parent.

How I can paginate using this structure?

**Comment by staudenmeir** [comment]
`Categorias::tree()->paginate(2)` is the base query for this kind of pagination. You also need to order the query to get consistent results. Your example looks like [`depthFirst()`](https://github.com/staudenmeir/laravel-adjacency-list?tab=readme-ov-file#order).

Use the [`depth` attribute](https://github.com/staudenmeir/laravel-adjacency-list?tab=readme-ov-file#depth) to determine how many dashes you need to display.

**Reply by ElisameAraujo** [reply]
Okay, I have a update, I workaround like this:

I have 2 variables that I pass to view, the first load the method `tree()->get()->toTree()` where I can use the method child in the view to load the childs of parent categories.

The second is the parent category itself, so... I have:

```
$categorias = Categorias::tree()->get()->toTree();
$categoriasMae = Categorias::where('categoriaMae', NULL)->paginate(10);
return view('admin.blog.categorias', compact('categorias','categoriasMae'));
```
And in my view:
```
@foreach ($categoriasMae as $categoria)
            <tr>
                <td>{{ $categoria->categoriaID }}</td>
                <td>{{ $categoria->categoriaNome }}</td>
                <td>{{ $categoria->posts->count() }}</td>
                <td>@if ($categoria->categoriaMae !== NULL)
                    @foreach ($categoria->subcategorias as $categoriaMae)
                        {{ $categoriaMae->CategoriaProduto }}
                    @endforeach
                @else
                    -
                @endif</td>
                <td>{{ $categoria->slugCategoria }}</td>
                <td>
                    <span class="btn-group">
                        <a href="#" class="btn btn-sm btn-outline-primary editar" data-bs-toggle="modal" data-bs-target="#editarCategoria" data-categoria="{{ $categoria->nomeCategoria }}" data-categoria-id="{{ $categoria->categoriaID }}">
                            <i class="fa-regular fa-pen-to-square" data-bs-toggle="tooltip" data-bs-placement="top" title="Editar"></i>
                        </a>
                        <a href="#" class="btn btn-sm btn-outline-danger excluir" data-bs-toggle="modal" data-bs-target="#excluirCategoria" data-categoria-excluir="{{ $categoria->nomeCategoria }}" data-cat-id="{{ $categoria->categoriaID }}">
                            <i class="fa-regular fa-trash-can" data-bs-toggle="tooltip" data-bs-placement="top" title="Excluir" ></i>
                        </a>
                    </span>
                </td>
            </tr>
            <tr>
                @foreach ($categoria->children as $subcategoria)
                    <td>{{ $subcategoria->categoriaID }}</td>
                    <td>&mdash; {{ $subcategoria->categoriaNome }}</td>
                    <td>{{ $subcategoria->posts->count() }}</td>
                    <td>{{ $categoria->categoriaNome }}</td>
                    <td>{{ $subcategoria->slugCategoria }}</td>
                    <td>
                        <span class="btn-group">
                            <a href="#" class="btn btn-sm btn-outline-primary editar" data-bs-toggle="modal" data-bs-target="#editarCategoria" data-categoria="{{ $categoria->nomeCategoria }}" data-categoria-id="{{ $categoria->categoriaID }}">
                                <i class="fa-regular fa-pen-to-square" data-bs-toggle="tooltip" data-bs-placement="top" title="Editar"></i>
                            </a>
                            <a href="#" class="btn btn-sm btn-outline-danger excluir" data-bs-toggle="modal" data-bs-target="#excluirCategoria" data-categoria-excluir="{{ $categoria->nomeCategoria }}" data-cat-id="{{ $categoria->categoriaID }}">
                                <i class="fa-regular fa-trash-can" data-bs-toggle="tooltip" data-bs-placement="top" title="Excluir" ></i>
                            </a>
                        </span>
                    </td>
                </tr>
                @endforeach
        @endforeach
```
So, using this way I could render the pagination using the variable `$categoriasMae` with Laravel links:
`{{ $categoriasMae->onEachSide(2)->withQueryString()->links('vendor/pagination/bs-5')}}`

And the result is:
![image](https://github.com/staudenmeir/laravel-adjacency-list/assets/40779799/cb1f3f80-8904-45e7-b4f9-e8d77d5f8571)

So, using the suggestions, I came with this solution.
Thanks for the help!
